### ISSUE TITLE
Go1.6Early: html/template: cannot redefine "_default/single.html" after it has executed

### ISSUE THREAD

**Issue by anthonyfok** [spf13/hugo#1470]
Just an early heads up of what is to come:

Since a day or two ago, the [Travis CI test](https://travis-ci.org/spf13/hugo/builds) for building Hugo with "Go tip", i.e. the development tip for the future Go 1.6, no longer passes, and instead failing with the errors when running `go test`:

    html/template: cannot redefine "_default/single.html" after it has executed

Example build that exhibits this error: https://travis-ci.org/spf13/hugo/jobs/83273687

It seems to have to do with [this recent commit](https://go.googlesource.com/go/+/12dfc3bee482f16263ce4673a0cce399127e2a0d) ([diff](https://go.googlesource.com/go/+/652d2386e99526eccc6634af2eaa7ff4c9a113c4%5E%21/)) to Go:

> text/template, html/template: add block keyword and permit template redefinition
>
> This change adds a new "block" keyword that permits the definition
of templates inline inside existing templates, and loosens the
restriction on template redefinition. Templates may now be redefined,
but in the html/template package they may only be redefined before
the template is executed (and therefore escaped).
>
> The intention is that such inline templates can be redefined by
subsequent template definitions, permitting a kind of template
"inheritance" or "overlay". (See the example for details.)
>
> Fixes [#3812](https://github.com/golang/go/issues/3812)

**Comment by bep** [comment]
It would be useful if you could spin up a Hugo site with Go 1.6 and see if same happens. If yes, I would say that this is a breaking change and a bug in Go.

**Comment by bep** [comment]
I  built my site with a 1.6 Hugo, and it works fine. It is some of our tests that needs fixing. Should be doable.

**Comment by anthonyfok** [comment]
I finally did some testing.  Hugo's own documentation would trigger the error, i.e. by running `hugo -s docs` from the root of the hugo Git repository.

I was also able to reproduce the error on my own yet-to-be-published blog.  I soon discovered that I had a zero-byte `content/a.md` that I had inadvertently left behind some months ago.

I was then able to reproduce the error on a minimal site like so (with `hugo` compiled with the go1.6 branch):

```
$ hugo new site testsite
$ cd testsite
$ touch content/a.md
$ touch content/b.md
$ hugo
```

Need to create at least two such empty files to trigger the error.  And indeed, no theme is needed.

The `hugo` run would also list available templates, among which `__a.html` and `__b.html` are also available.  That `__` prefix led me to the following snippet in `pageRenderer()` in `hugolib/site.go` (https://github.com/spf13/hugo/blob/master/hugolib/site.go#L938-L961):

```go
		if !p.IsRenderable() {
			self := "__" + p.TargetPath()
			_, err := s.Tmpl.New(self).Parse(string(p.Content))
```

Since `fmt.Errorf("html/template: cannot redefine %q after it has executed", name)` comes from `go/src/html/template/template.go` (https://github.com/golang/go/blob/master/src/html/template/template.go#L180), this looks like the suspect.

And indeed, when I add a `continue` to skip `_, err := s.Tmpl.New(self).Parse(string(p.Content))` altogether, the `html/template: cannot redefine XXXXXX after it has executed` error goes away.

Why would Hugo try to create a template out of a unrenderable file?  That did not make sense to me, so I used a series of `git blame` to trace the "culprit", and eventually found this:

```
commit 311e10222301d47422f970c460383879ad78f681
Author: Noah Campbell <noahcampbell@gmail.com>
Date:   Wed Sep 18 14:21:27 2013 -0700

    Allow non-markdown content in content directory

    Allow content that is not markdown and does not need to be rendered to
    exists in the content directory.  Currently any valid html or xml
    document can exist.  Templates are applied to these documents as well.
    If you need to have content that doesn't have templates or AbsUrlify
    like operations, then continue to put this content in static and it will
    be copied over.
```

"Templates are applied to these documents as well."...  Okay... Hmm...

What about the addition of this new error message `html/template: cannot redefine...` in Go1.6?

```
commit 12dfc3bee482f16263ce4673a0cce399127e2a0d
Author: Andrew Gerrand <adg@golang.org>
Date:   Fri Aug 28 15:31:51 2015 +1000

    text/template, html/template: add block keyword and permit template redefinition
    
    This change adds a new "block" keyword that permits the definition
    of templates inline inside existing templates, and loosens the
    restriction on template redefinition. Templates may now be redefined,
    but in the html/template package they may only be redefined before
    the template is executed (and therefore escaped).
    
    The intention is that such inline templates can be redefined by
    subsequent template definitions, permitting a kind of template
    "inheritance" or "overlay". (See the example for details.)
    
    Fixes #3812
    
    Change-Id: I733cb5332c1c201c235f759cc64333462e70dc27
    Reviewed-on: https://go-review.googlesource.com/14005
    Reviewed-by: Rob Pike <r@golang.org>
```

Now, what intrigues me is that in the minimal site example above, the `html/template: cannot redefine "__a.html" after it has executed` is not because `s.Tmpl.New("__a.html").Parse(string(p.Content))` is called twice—hugo only makes the call once.  The error actually happens during the call to `s.Tmpl.New("__b.html").Parse(string(p.Content))`!  How on earth would the definition of template `__b.html` lead to the redefinition of template `__a.html`?

As it turns out, it is Go's html/template’s `Parse()` own doing (https://github.com/golang/go/blob/master/src/html/template/template.go#L169-L186):

```go
	// In general, all the named templates might have changed underfoot.
	// Regardless, some new ones may have been defined.
	// The template.Template set has been updated; update ours.
	t.nameSpace.mu.Lock()
	defer t.nameSpace.mu.Unlock()
	for _, v := range ret.Templates() {
		name := v.Name()
		tmpl := t.set[name]
		if tmpl == nil {
			tmpl = t.new(name)
		} else if tmpl.escapeErr != nil {
			return nil, fmt.Errorf("html/template: cannot redefine %q after it has executed", name)
		}
		// Restore our record of this text/template to its unescaped original state.
		tmpl.escapeErr = nil
		tmpl.text = v
		tmpl.Tree = v.Tree
	}
```

So, why does `Parse()` go through all the previously defined templates and `set` them again?  I know too little about Go's template system to be able to answer that.

----

Curious to find what the actual `escapeErr` it is, I decided to print it out:

```
_internal/shortcodes/ref.html: escapeErr = <nil>
_internal/shortcodes/relref.html: escapeErr = <nil>
_internal/shortcodes/highlight.html: escapeErr = <nil>
...
__a.html: escapeErr = template escaped correctly

(then comes the error when creating "__b.html" template: html/template: cannot redefine "__a.html" after it has executed)
```

What!?  "template escaped correctly" is the escapeErr?  That's <s>weird</s> interesting...

From `type template struct { }` definition (https://github.com/golang/go/blob/master/src/html/template/template.go#L19-L29):

```go
	// Sticky error if escaping fails, or escapeOK if succeeded.
	escapeErr error
```

And from https://github.com/golang/go/blob/master/src/html/template/template.go#L31-L32:

```go
// escapeOK is a sentinel value used to indicate valid escaping.
var escapeOK = fmt.Errorf("template escaped correctly")
```

Then I found it used in a few places, e.g., in `lookupAndEscapeTemplate()`, like so:

```go
	if tmpl.escapeErr != nil && tmpl.escapeErr != escapeOK {
		return nil, tmpl.escapeErr
	}
```

But not so in `Parse()`—`escapeErr` is currently not checked against `escapeOK` there:

```go
		} else if tmpl.escapeErr != nil {
			return nil, fmt.Errorf("html/template: cannot redefine %q after it has executed", name)
		}
```

And the following patch seems to have squelched the error with `hugo -s docs` and `go test ./...` etc.:

```diff
--- a/src/html/template/template.go
+++ b/src/html/template/template.go
@@ -176,7 +176,7 @@ func (t *Template) Parse(src string) (*Template, error) {
 		tmpl := t.set[name]
 		if tmpl == nil {
 			tmpl = t.new(name)
-		} else if tmpl.escapeErr != nil {
+		} else if tmpl.escapeErr != nil && tmpl.escapeErr != escapeOK {
 			return nil, fmt.Errorf("html/template: cannot redefine %q after it has executed", name)
 		}
 		// Restore our record of this text/template to its unescaped original state.
```

So, is that it?  A possible bug in the current Go1.6 branch?

**Comment by anthonyfok** [comment]
Sorry for the very wordy and long message.  My puny little brain was exploding and needed to do a "brain dump", hehe.

**Comment by bep** [comment]
@anthonyfok I have looked into this myself and didn't get too bright about it. But I still don't understand the ` _, err := s.Tmpl.New(self).Parse(string(p.Content))` and `Page.IsRenderable` logic.

I would says this looks like a bug in Go, at least in the sense in that this is a pretty breaking change; but if we could fix this by removing functionality we just don't use (or understand.)

**Comment by anthonyfok** [comment]
@bep Thanks for the note.  Just a quick update: my proposed patch to `go/src/html/template/template.go` does not fly: It fails at go/src/html/template/clone_test.go:92.  The "template escaped correctly" is set intentionally after an ExecuteTemplate() call, and Go1.6 uses that to test for "redefinition after execute".

Another note: `hugo -s docs` fails because of two empty files `docs/content/indexes/category.md` and `docs/content/taxonomies/category.md` which somehow did not get deleted (commit a87f171) but got left behind as zero-length files on my computer.  `hugo -s docs` run fine from a fresh `git clone https://github.com/spf13/hugo.git`.

> But I still don't understand the ` _, err := s.Tmpl.New(self).Parse(string(p.Content))` and `Page.IsRenderable` logic.

Me neither.  @noahcampbell, could you please enlighten us?  :-)

**Comment by noahcampbell** [comment]
2013 was a long time ago :).  I'm not sure I remember my intent.

**Comment by moorereason** [comment]
My hunch is that this is not a Go bug.  The `_, err := s.Tmpl.New(self).Parse(string(p.Content))` line ([hugolib/site.go:L958](https://github.com/spf13/hugo/blob/e29ede70556c2b5207032d30351f3b53a298ca0c/hugolib/site.go#L958)) is the same one that's triggering a race condition in Go 1.5 (see #1601).

@natefinch is the one responsible for those changes in `pageRenderer`.

**Comment by moorereason** [comment]
This issue is a dupe of #1601.

**Comment by bep** [comment]
Closing as a duplicate.

**Comment by bep** [comment]
It's still an issue after the fix for #1601.

**Comment by moorereason** [comment]
See my https://github.com/spf13/hugo/issues/1601#issuecomment-165349295.

I think the root cause of this issue and #1601 is that we are rendering into and reusing a **single Template instance** (`Site.Tmpl`).  PR #1717 only solved the race condition because you removed concurrency in the rendering process.

I haven't messed with Go 1.6, but here's what Andrew said in the Go commit message:  "Templates may now be redefined, but in the html/template package they may only be redefined before the template is executed (and therefore escaped)."  See [Go Change 14005](https://go-review.googlesource.com/#/c/14005/).  Again, I think once the first template is defined and executed, you can't redefine that template instance in Go 1.6.

My suggestion is to return to using concurrency in `pageRenderer` (ie. revert part of #1717) and creating a new Template instance for each page.  I haven't had time to work on a fix for this issue or #1601, but I would hope that removing the single template instance would solve both problems.

**Comment by bep** [comment]
I have a fix to this issue. Will commit in a few minutes.
